src/Compilers/CSharp/Test/Syntax/Parsing/ParsingErrorRecoveryTests.cs

```diff
@@ -6274,6 +6274,46 @@ class C
             Assert.Equal((int)ErrorCode.ERR_IdentifierExpected, file.Errors()[0].Code);
         }
 
+        [Fact]
+        public void TestMissingGreaterThanTokenInReturnType()
+        {
+            var text = @"class Class
+{
+    List<int M(string a)
+    {
+        var x = 42;
+    }
+}";
+            SyntaxTree syntaxTree = SyntaxFactory.ParseSyntaxTree(text);
+            Assert.Equal(text, syntaxTree.GetCompilationUnitRoot().ToFullString());
+            
+            Assert.Equal(new [] { "(3,14): error CS1003: Syntax error, '>' expected" },
+                syntaxTree.GetDiagnostics()
+                    .Select(d => ((IFormattable)d).ToString(null, EnsureEnglishUICulture.PreferredOrNull)));
+
+            Assert.Equal(16, syntaxTree.FindNodeOrTokenByKind(SyntaxKind.MethodDeclaration).Position);
+        }
+
+        [Fact]
+        public void TestMissingCommaTokenInReturnType()
+        {
+            var text = @"class Class
+{
+    Dictionary<int string> M(string a)
+    {
+        var x = 42;
+    }
+}";
+            SyntaxTree syntaxTree = SyntaxFactory.ParseSyntaxTree(text);
+            Assert.Equal(text, syntaxTree.GetCompilationUnitRoot().ToFullString());
+
+            Assert.Equal(new[] { "(3,20): error CS1003: Syntax error, ',' expected" },
+                syntaxTree.GetDiagnostics()
+                    .Select(d => ((IFormattable)d).ToString(null, EnsureEnglishUICulture.PreferredOrNull)));
+
+            Assert.Equal(16, syntaxTree.FindNodeOrTokenByKind(SyntaxKind.MethodDeclaration).Position);
+        }
```

Comment:
we need a lot more tests of this new behavior.  especially where tehre might be a `>` but we now assume that the lack of a `,` causes us to terminate the list earlier.I would particularly like a PR that first adds those tests (and has the errors currently), then makes the parser change, and has to update those tests.

i.e. to see if this causes a regressive behavior in some cases.

Reply:
also, we should have tests that validate the actual shape of the tree.  See several other parsing test files for examples of this.  For example, nothing about this test helps me know if the parser really understood this was a method with a generic return type.

Reply:
@CyrusNajmabadi
That's fair enough and I will attempt to do so.
I think, but have not yet verified that error recovery here, although not perfect, is absolutely better than it was. i.e.there are no cases where error recovery is worse than it was previously.

Reply:
Excluding a case where two commas in the type argument list are missed, but that is rare enough that I will not worry about it too much